### PR TITLE
Maded Title-Objects directy sendable to targets

### PR DESCRIPTION
**Pull request by TheFruxz**
Hey!

This is my first submission of a change to PaperMC/Paper (API).

I myself think that title objects should also be sent by themselves (like in ParticleBuilder). Therefore I made this suggestion for improvement.

With kind regards
Fruxz (@TheFruxz)

### CODE REVIEW

File: Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
++    /**
++     * Sends the title directly to the defined players
++     *
++     * <p>The players can be null, but the array itself not</p>
++     *
++     * @param players defines the targets of the title
++     */
++    public void send(@NotNull Player[] players) {
++        if (players != null) {
++            Arrays.stream(players).forEach((player -> {
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch:62]: Awful use of a stream. Just use a good ol' for loop.

File: Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
++
++    /**
++     * Sends the title directly to the defined players
++     *
++     * <p>The players can be null, but the list itself not</p>
++     *
++     * @param players defines the targets of the title
++     */
++    public void send(@NotNull Collection<Player> players) {
++        if (players != null) {
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch:44]: Don't null check. If the user passes null, it's their problem and they're the ones who have to fix their NPE.

File: Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
++
++    /**
++     * Sends the title directly to the defined players
++     *
++     * <p>The players can be null, but the array itself not</p>
++     *
++     * @param players defines the targets of the title
++     */
++    public void send(@NotNull Player[] players) {
++        if (players != null) {
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch:61]: Don't null check. If the user passes null, it's their problem and they're the ones who have to fix their NPE.

File: Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+@@ -224,6 +227,51 @@ public final class Title {
+         return this.fadeOut;
+     }
+ 
++    /**
++     * Sends the title directly to an player
++     *
++     * <p>The player value can not be null!</p>
++     *
++     * @param player defines the target of the title
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch:30]:
```suggestion
+     * @param player the receiver of the title
```

File: Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
++    public void send(@NotNull Player player) {
++        player.sendTitle(this);
++    }
++
++    /**
++     * Sends the title directly to the defined players
++     *
++     * <p>The players can be null, but the list itself not</p>
++     *
++     * @param players defines the targets of the title
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch:41]:
```suggestion
+     * @param players the receivers of the title
```

File: Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
++            }));
++        }
++    }
++
++    /**
++     * Sends the title directly to the defined players
++     *
++     * <p>The players can be null, but the array itself not</p>
++     *
++     * @param players defines the targets of the title
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch:58]:
```suggestion
+     * @param players the receivers of the title
```

File: Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
++                    send(player);
++                }
++            }));
++        }
++    }
++
++    /**
++     * Sends the title directly to the defined players
++     *
++     * <p>The players can be null, but the array itself not</p>
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch:56]:
```suggestion
```

File: Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
++     * @param player defines the target of the title
++     */
++    public void send(@NotNull Player player) {
++        player.sendTitle(this);
++    }
++
++    /**
++     * Sends the title directly to the defined players
++     *
++     * <p>The players can be null, but the list itself not</p>
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch:39]:
```suggestion
```

File: Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
++     *
++     * @param players defines the targets of the title
++     */
++    public void send(@NotNull Collection<Player> players) {
++        if (players != null) {
++            players.forEach((player -> {
++                if (player != null && player.isOnline()) {
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch:46]: Don't null check. If the user passes null, it's their problem and they're the ones who have to fix their NPE.

Checking online is also likely unnecessary; if the player is offline, the plugin shouldn't pass them in the first place.

File: Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
++     * Sends the title directly to the defined players
++     *
++     * <p>The players can be null, but the array itself not</p>
++     *
++     * @param players defines the targets of the title
++     */
++    public void send(@NotNull Player[] players) {
++        if (players != null) {
++            Arrays.stream(players).forEach((player -> {
++                if (player != null && player.isOnline()) {
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch:63]: Don't null check. If the user passes null, it's their problem and they're the ones who have to fix their NPE.

Checking online is also likely unnecessary; if the player is offline, the plugin shouldn't pass them in the first place.

File: Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+ import static com.google.common.base.Preconditions.checkNotNull;
+ import static com.google.common.base.Preconditions.checkState;
+@@ -224,6 +227,51 @@ public final class Title {
+         return this.fadeOut;
+     }
+ 
++    /**
++     * Sends the title directly to an player
++     *
++     * <p>The player value can not be null!</p>
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Maded-Title-objects-directly-sendable.patch:28]:
```suggestion
```

**Review by Proximyst — CHANGES_REQUESTED** [review]
Good first contribution, just some nitpicks then it'd be good to go.

Also read the contributing file to read how to fixup this into the patch which makes the class. This doesn't need to be a new patch.

**Comment by TheFruxz** [comment]
Hey, many thanks for the help with my first PR! I tried to commit all the changes to the repository like in the [documentation](https://github.com/PaperMC/Paper/blob/master/CONTRIBUTING.md#automatic-method) and integrated the commit directly into this PR. Thanks again for the help!

**Comment by TheFruxz** [comment]
> Hey, many thanks for the help with my first PR! I tried to commit all the changes to the repository, and integrated the commit directly into this PR. Thanks again for the help!

P.S. My IDE apparently re-sorted the this() objects in fixup, but did not make any changes. IDK Why

**Comment by TheFruxz** [comment]
For this feature, I created a new PR and I closing this PR to to avoid confusion and to fix the 'unknown repository' GitHub-Issue

NEW ONE: #4648 

Thanks for helping @Proximyst!

**Pull request closed without merging**